File: src/utils/exchange.ts
import type { bills, exchange as exchangeType } from "../types"
import { initialExchange } from "../constants"

export const exchangeQuinary = (amount: number): exchangeType => {
    const exchanges: exchangeType = initialExchange;

    const totalFives = Math.floor(amount / 5);

    exchanges[25] = Math.floor(totalFives / 5);
    exchanges[10] = Math.floor((totalFives % 5) / 2);
    exchanges[5]  = (totalFives % 5) % 2;

    exchanges[1]  = amount % 5;

    return exchanges;
}

export const exchangeGreedy = (amount: number, bills: bills[]): exchangeType => {
    const exchanges: exchangeType = initialExchange;

    for (const bill of bills) {
        exchanges[bill] = Math.floor(amount / bill);
        amount %= bill;
    }

    exchanges[1] = amount;

    return exchanges;
}

File: src/utils/exchange.cpp
#include <iostream>
#include <vector>
using namespace std;

struct Exchange {
    int b25;
    int b10;
    int b5;
    int b1;
};

Exchange exchangeQuinary(int amount) {
    Exchange ex = {0, 0, 0, 0};

    int totalFives = amount / 5;

    ex.b25 = totalFives / 5;
    ex.b10 = (totalFives % 5) / 2;
    ex.b5  = (totalFives % 5) % 2;

    ex.b1  = amount % 5;

    return ex;
}

Exchange exchangeGreedy(int amount, const vector<int>& bills) {
    Exchange ex = {0, 0, 0, 0};

    for (int bill : bills) {
        int count = amount / bill;
        amount %= bill;

        if (bill == 25) ex.b25 = count;
        else if (bill == 10) ex.b10 = count;
        else if (bill == 5)  ex.b5  = count;
        else if (bill == 1)  ex.b1  = count;
    }

    ex.b1 += amount;
    return ex;
}